### PR TITLE
Fix sporadic IOException on pooled connection server-induced shutdown

### LINKED ISSUE #13: The connection taken from pool which was shut down by the server leads to IOException
Hi.
First, I would like to thank you for creating and maintaining this library. The default .NET Framework `HttpClient` implementation based on `HttpWebRequest` is really lacking in terms of stability (especially timeout/cancellation handling), and your library is a great fix for that.

Unfortunately, since we started using it, we encountered the following issue. With some servers which shut down the pooled connections from their own side, the request that follows fails with HttpRequestException wrapping an IOException:
```
System.Net.Http.HttpRequestException: An error occurred while sending the request.
async Task<HttpResponseMessage> System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken) in StandardSocketsHttpHandler/StandardSocketsHttpHandler/Net/Http/SocketsHttpHandler/HttpConnection.cs:line 724
async Task<HttpResponseMessage> System.Net.Http.HttpConnectionPool.SendWithNtConnectionAuthAsync(HttpConnection connection, HttpRequestMessage request, bool doRequestAuth, CancellationToken cancellationToken) in StandardSocketsHttpHandler/StandardSocketsHttpHandler/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:line 316
async Task<HttpResponseMessage> System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, bool doRequestAuth, CancellationToken cancellationToken) in StandardSocketsHttpHandler/StandardSocketsHttpHandler/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:line 296
...
---> System.IO.IOException: The server returned an invalid or unrecognized response.
async Task<HttpResponseMessage> System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken) in StandardSocketsHttpHandler/StandardSocketsHttpHandler/Net/Http/SocketsHttpHandler/HttpConnection.cs:line 516
```

I believe that this happens due to the [`ConsumeReadAheadTask`](https://github.com/TalAloni/StandardSocketsHttpHandler/blob/2.2/StandardSocketsHttpHandler/Net/Http/SocketsHttpHandler/HttpConnection.cs#L508) reading zero bytes. In the original SocketsHttpHandler this [was addressed](https://github.com/dotnet/runtime/commit/288bfa00fe28b4b25637c362683dd8c6b6573351#diff-595f700a4c83b735d856fe14b5ebe8c7f7c18ae6e586adc8b398a997a0678274). While I understand that you're probably not going to integrate all the "new" changes to SocketsHttpHandler, this one routinely breaks conversations with remote servers and probably is a good candidate for a backport.

I guess we can work around the issue by making the connection pool lifetime smaller for this specific service, but I don't think that will protect us from the problem in all cases.

Thanks for taking a look.

### PR DESCRIPTION
**Pull request by onyxmaster**
Resolves #13.

### CODE REVIEW

**Comment by TalAloni** [comment]
In the [original PR](https://github.com/dotnet/runtime/commit/288bfa00fe28b4b25637c362683dd8c6b6573351#diff-595f700a4c83b735d856fe14b5ebe8c7f7c18ae6e586adc8b398a997a0678274) to Microsoft at least one test was added ( HttpClientHandlerTest_RequestRetry ) but attempting to copy and run this test on your branch - the test fails - what am I missing?

**Comment by onyxmaster** [comment]
Thanks for noticing this problem, I'll take a look later this week.

**Comment by onyxmaster** [comment]
The tests that were added as the part of the original PR test different aspects of modified retry mechanism, most of which are not included in the proposed PR (unfortunately the original PR fixes many things in different places, while I chose to fix only one issue). The test that is closest to testing the fixed issue is [GetAsyncOnExistingConnection_RetryOnConnectionClosed_Success](https://github.com/dotnet/runtime/blob/288bfa00fe28b4b25637c362683dd8c6b6573351/src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.RequestRetry.cs#L49), but unchanged it won't show any problems with pre-fix code since it only deals with `GET` HTTP method, while the issue with connection reuse manifests itself only when there is some posted content.
I adapted the test, renaming it to [PostAsyncOnExistingConnection_RetryOnConnectionClosed_Success](https://github.com/TalAloni/StandardSocketsHttpHandler/pull/15/commits/d88df7999423dad95acfe767368f5aa0d4b1707c), converting `GET` to `POST`. With this test in place, reverting the [4aa2ee](https://github.com/TalAloni/StandardSocketsHttpHandler/pull/15/commits/4aa22ee4ee9169166dfc4da6ac1ed1daceea18d5) commit fails the test, while leaving it applied indicates that the proposed fix is, in fact, working.

**Comment by TalAloni** [comment]
**Edit:**
I believe the appropriate file for the test is 'HttpRetryProtocolTests.cs'

**Comment by onyxmaster** [comment]
The test I added does not pass before the applied fix:
```
System.Net.Http.Functional.Tests.SocketsHttpHandler_HttpClientHandlerTest.PostAsyncOnExistingConnection_RetryOnConnectionClosed_Success
   Source: HttpClientHandlerTest.cs line 2229
   Duration: 3,4 sec

  Message: 
System.Net.Http.HttpRequestException : An error occurred while sending the request.
---- System.IO.IOException : The server returned an invalid or unrecognized response.

  Stack Trace: 
<SendAsyncCore>d__61.MoveNext() line 719
--- End of stack trace from previous location where exception was thrown ---
TaskAwaiter.ThrowForNonSuccess(Task task)
TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
ConfiguredTaskAwaiter.GetResult()
<SendWithNtConnectionAuthAsync>d__40.MoveNext() line 316
--- End of stack trace from previous location where exception was thrown ---
TaskAwaiter.ThrowForNonSuccess(Task task)
TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
ConfiguredTaskAwaiter.GetResult()
<SendWithRetryAsync>d__39.MoveNext() line 296
--- End of stack trace from previous location where exception was thrown ---
TaskAwaiter.ThrowForNonSuccess(Task task)
TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
ConfiguredTaskAwaiter.GetResult()
<SendAsync>d__4.MoveNext() line 32
--- End of stack trace from previous location where exception was thrown ---
TaskAwaiter.ThrowForNonSuccess(Task task)
TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
TaskAwaiter`1.GetResult()
<<PostAsyncOnExistingConnection_RetryOnConnectionClosed_Success>b__87_0>d.MoveNext() line 2242
--- End of stack trace from previous location where exception was thrown ---
TaskAwaiter.ThrowForNonSuccess(Task task)
TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
ConfiguredTaskAwaiter.GetResult()
<WhenAllOrAnyFailed>d__3.MoveNext() line 64
--- End of stack trace from previous location where exception was thrown ---
ExceptionDispatchInfo.Throw()
<WhenAllOrAnyFailed>d__3.MoveNext() line 92
--- End of stack trace from previous location where exception was thrown ---
TaskAwaiter.ThrowForNonSuccess(Task task)
TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
TaskAwaiter.GetResult()
<<CreateClientAndServerAsync>b__0>d.MoveNext() line 84
--- End of stack trace from previous location where exception was thrown ---
TaskAwaiter.ThrowForNonSuccess(Task task)
TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
TaskAwaiter.GetResult()
<CreateServerAsync>d__14.MoveNext() line 67
--- End of stack trace from previous location where exception was thrown ---
TaskAwaiter.ThrowForNonSuccess(Task task)
TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
TaskAwaiter.GetResult()
<PostAsyncOnExistingConnection_RetryOnConnectionClosed_Success>d__87.MoveNext() line 2231
--- End of stack trace from previous location where exception was thrown ---
TaskAwaiter.ThrowForNonSuccess(Task task)
TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
TaskAwaiter.ThrowForNonSuccess(Task task)
TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
TaskAwaiter.ThrowForNonSuccess(Task task)
TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
----- Inner Stack Trace -----
<SendAsyncCore>d__61.MoveNext() line 516
```

**Comment by onyxmaster** [comment]
>I believe the appropriate file for the test is 'HttpRetryProtocolTests.cs'

I explicitly chose not to add new file, since the original PR both split the HttpClientHandlerTest into several parts and added new methods there. If you believe i should split the file -- I've got no problems with that, let me improve it.

File: StandardSocketsHttpHandler/Net/Http/SocketsHttpHandler/HttpConnection.cs [hunk 1]
```diff
@@ -363,8 +364,8 @@ public async Task<HttpResponseMessage> SendAsyncCore(HttpRequestMessage request,
             HttpMethod normalizedMethod = HttpMethodUtils.Normalize(request.Method);
             bool hasExpectContinueHeader = request.HasHeaders() && request.Headers.ExpectContinue == true;
 
-            Debug.Assert(!_canRetry);
             _canRetry = true;
```
Comment (TalAloni) [inline StandardSocketsHttpHandler/Net/Http/SocketsHttpHandler/HttpConnection.cs:367]: In the original PR _canRetry was set here to false, In your PR it remains true, is that on purpose?
After your PR _canRetry will always be true

Comment (onyxmaster) [reply StandardSocketsHttpHandler/Net/Http/SocketsHttpHandler/HttpConnection.cs:367]: Thanks for noticing this, yes, this looks like a bug. I'll take a look at it later.

**Comment by TalAloni** [comment]
I've noticed that HttpConnectionPool.SendWithRetryAsync was not modified and isNewConnection is still being checked.

**Comment by onyxmaster** [comment]
Sorry for the long delay and thanks for paying a lot of attention to the issue, but it appears that this PR won't accomplish anything if the content is being already sent. I ran the test introduced above vs the SocketsHttpHandler in the `release/6.0` branch of .NET Runtime, and it failed. Looks there is nothing to backport, so I think I should close this PR and revert back to pool timeout setup.

**Pull request closed without merging**